### PR TITLE
Remove webroot fallback to approot

### PR DESCRIPTION
**Pull request by pakrym**
Issue: https://github.com/aspnet/Hosting/issues/480
@Tratcher @davidfowl

### CODE REVIEW

**Comment by Tratcher** [comment]
@blowdart

File: src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using Microsoft.AspNet.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    internal class NullFileProvider : IFileProvider
+    {
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            return new NullDirectoryContents();
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs]: You don't need to create a new one every time.

File: src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs [hunk 2]
```diff
@@ -91,7 +91,10 @@ public static bool IsProduction(this IHostingEnvironment hostingEnvironment)
             {
                 throw new ArgumentNullException(nameof(hostingEnvironment));
             }
-
+            if (hostingEnvironment.WebRootPath == null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: We need to get rid of MapPath. It's a little misleading

File: src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs]: Not that it matters :smile:

**Comment by blowdart** [comment]
I love you all so very much. :shipit:

File: src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs (hunk 2 above)
Comment (pakrym) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: In this pr?

File: src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs (hunk 1 above)
Comment (pakrym) [inline src/Microsoft.AspNet.Hosting/Internal/NullFileProvider.cs]: Yes, this and NullChangeToken can be singletons, but do we care?

File: test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        [Fact]
+        public void DefaultsToWwwrootSubdir()
+        {
+            var env = new HostingEnvironment();
+
+            env.Initialize("testroot", new WebHostOptions(), null);
+
+            Assert.Equal(Path.GetFullPath($"testroot{Path.DirectorySeparatorChar}wwwroot"), env.WebRootPath);
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs]: `Path.Combine()`

File: test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            Assert.IsAssignableFrom<PhysicalFileProvider>(env.WebRootFileProvider);
+        }
+
+        [Fact]
+        public void DefaultsToNullFileProvider()
+        {
+            var env = new HostingEnvironment();
+
+            env.Initialize($"testroot{Path.DirectorySeparatorChar}wwwroot", new WebHostOptions(), null);
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs]: `Path.Combine()`

File: test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/TextFile.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+﻿A text file.
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/TextFile.txt]: Where is this file used?

File: test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/HostingEnvironmentExtensionsTests.cs]: sort usings

File: test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/TextFile.txt (hunk 5 above)
Comment (pakrym) [inline test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/TextFile.txt]: Too keep `wwwroot` directory

Comment (cesarbs) [inline test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/TextFile.txt]: Change its contents to indicate that is the intention of having that file there.

**Comment by pakrym** [comment]
:up: :date:

File: src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs (hunk 2 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: Nah

Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: string.IsNullOrEmpty

Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: Also the casing of webroot is strange.

**Comment by davidfowl** [comment]
Minor comments then:shipit:

File: src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs [hunk 7]
```diff
@@ -91,7 +91,10 @@ public static bool IsProduction(this IHostingEnvironment hostingEnvironment)
             {
                 throw new ArgumentNullException(nameof(hostingEnvironment));
             }
-
+            if (string.IsNullOrEmpty(hostingEnvironment.WebRootPath))
+            {
+                throw new InvalidOperationException("Can not map path because webroot path is not set");
```
Comment (victorhurdugaci) [inline src/Microsoft.AspNet.Hosting.Abstractions/HostingEnvironmentExtensions.cs]: Can not => Cannot

File: test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/README [hunk 8]
```diff
@@ -0,0 +1 @@
+﻿This file is here to keep directories needed by tests. Do no remove it.
```
Comment (victorhurdugaci) [inline test/Microsoft.AspNet.Hosting.Tests/testroot/wwwroot/README]: "not" :)

**Comment by davidfowl** [comment]
Make an announcement

**Pull request merged**